ctru-rs/src/gfx.rs

```diff
@@ -48,14 +48,14 @@ impl Gfx {
     }
 
     /// Enable or disable the 3D stereoscopic effect
-    pub fn set_3d_enabled(&self, enabled: bool) {
+    pub fn set_3d_enabled(&mut self, enabled: bool) {
         unsafe { ctru_sys::gfxSet3D(enabled) }
     }
 
     /// Enable or disable the wide screen mode (top screen).
     ///
     /// This only works when 3D is disabled.
-    pub fn set_wide_mode(&self, enabled: bool) {
+    pub fn set_wide_mode(&mut self, enabled: bool) {
```

Comment:
This change means the wide mode example needs to be updated.

Reply:
The wide mode example might need to recreate the console when switching modes, if my understanding of the goal state is correct.

If you try the simple fix of making `gfx` mutable in the wide mode example to fix the compile error, it works. But we also don't want it to work since we're calling a mutable method on gfx after passing an immutable reference to console. The reason this works is due to [Non-Lexical Lifetimes](https://rust-lang.github.io/rfcs/2094-nll.html) (console's use of the gfx lifetime ends right after its creation). You can see this in action if you use the console at the end of the program, such as by adding `drop(_console)` (making the mutable gfx call in between two console usages). We should decide if this is something we want to prevent, since it is probably common for the user to never interact with the console after creation.

Reply:
Maybe we could have the console consume the Gfx object (pass ownership), and then expose a `fn gfx(&self) -> &Gfx` method on `Console`.

This might make some other use cases harder though, for example initializing two screens. I think I need to better understand how libctru intends its gfx API to work in some of these scenarios. Then we could find a way to enforce that usage pattern in Rust.

Reply:
No, the consuming isn't an option. The user wants to keep ownership at all times of the Gfx service. Isn't there a macro, or a #[] statement to prevent this behaviour?

Reply:
Macros just generate code, so they don't fundamentally affect how Rust works (i.e. they can't turn off NLL, but they might be able to emulate turning it off by inserting a `drop(console)` call). I don't think we want to introduce any new macros or proc-macro attributes into user code at this point though.

I tried out the gfx mutable fix in the wide mode example again (make `gfx` var mutable) and it fails with this error, which is what we want to happen:
```
error[E0502]: cannot borrow `gfx` as mutable because it is also borrowed as immutable
  --> ctru-rs/examples/gfx_wide_mode.rs:26:13
   |
14 |     let _console = Console::init(&gfx, Screen::Top);
   |                                  ---- immutable borrow occurs here
...
26 |             gfx.set_wide_mode(!gfx.get_wide_mode());
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
...
33 | }
   | - immutable borrow might be used here, when `_console` is dropped and runs the `Drop` code for type `ctru::console::Console`
```

I guess I forgot to update my comment before, because I definitely tried this out. So it is working as we expect, but only because we have the `Drop` impl on `Console`.

To fix the wide screen example I tried something like this, but it doesn't compile...
```diff
diff --git a/ctru-rs/examples/gfx_wide_mode.rs b/ctru-rs/examples/gfx_wide_mode.rs
index f4ba365..64670c9 100644
--- a/ctru-rs/examples/gfx_wide_mode.rs
+++ b/ctru-rs/examples/gfx_wide_mode.rs
@@ -10,10 +10,10 @@ fn main() {
     ctru::init();
     let apt = Apt::init().unwrap();
     let hid = Hid::init().unwrap();
-    let gfx = Gfx::default();
-    let _console = Console::init(&gfx, Screen::Top);
+    let mut gfx = Gfx::default();
+    let mut console = Console::init(&gfx, Screen::Top);
 
-    println!("Press A to enable/disable wide screen mode.");
+    println!("Press A to enable/disable wide screen mode. Extra words here so the text wraps.");
 
     while apt.main_loop() {
         hid.scan_input();
@@ -23,7 +23,14 @@ fn main() {
         }
 
         if hid.keys_down().contains(KeyPad::KEY_A) {
+            drop(console);
+
             gfx.set_wide_mode(!gfx.get_wide_mode());
+
+            console = Console::init(&gfx, Screen::Top);
+            println!(
+                "Press A to enable/disable wide screen mode. Extra words here so the text wraps."
+            );
         }
 
         gfx.flush_buffers();
```

Reply:
I'm pretty sure this is a bug in rustc, so I opened an issue while we try to work around it:
https://github.com/rust-lang/rust/issues/92984

Reply:
Oh wow, I don't know if it's a problem in rustc, but for now I'd say we should just work on having an acceptable API (with the Screens and all). After that, we should work on testing it out.

Reply:
@Meziu @ian-h-chamberlain FYI I just learned the above rustc bug is now fixed on stable!